src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/4.diff.sql

```diff
@@ -83,5 +124,121 @@ AS
+--
+-- PARAMETERS
+--     @tagPath
+--         * The TagPath for the extended query tag for which we retrieve error(s).
+/***************************************************************************************/
+CREATE OR ALTER PROCEDURE dbo.GetExtendedQueryTagErrors (@tagPath VARCHAR(64))
+AS
+	SET NOCOUNT     ON
+	SET XACT_ABORT  ON
+	BEGIN TRANSACTION
+
+
+    DECLARE @tagKey INT        
+
+	SELECT @tagKey = TagKey
+	FROM dbo.ExtendedQueryTag WITH(XLOCK)
+	WHERE dbo.ExtendedQueryTag.TagPath = @tagPath
+
+	-- Check existence
+	IF (@@ROWCOUNT = 0)
+		THROW 50404, 'extended query tag not found', 1 
+	ELSE
+		SELECT * FROM dbo.ExtendedQueryTagError
+		WHERE TagKey = @tagKey
+
+	COMMIT TRANSACTION
+GO
+
+/***************************************************************************************/
+-- STORED PROCEDURE
+--    DeleteExtendedQueryTagErrors
+--
+-- DESCRIPTION
+--    Delete errors for specified extended query tag
+--
+-- PARAMETERS
+--     @tagPath
+--         * The extended query tag path
+/***************************************************************************************/
+CREATE OR ALTER PROCEDURE dbo.DeleteExtendedQueryTagErrors (@tagPath VARCHAR(64))
+AS
+	SET NOCOUNT     ON
+	SET XACT_ABORT  ON
+	BEGIN TRANSACTION
+
+    DECLARE @tagKey INT        
+
+	SELECT @tagKey = TagKey
+	FROM dbo.ExtendedQueryTag WITH(XLOCK)
+	WHERE dbo.ExtendedQueryTag.TagPath = @tagPath
+
+	-- Check existence
+	IF (@@ROWCOUNT = 0)
+		THROW 50404, 'extended query tag not found', 1 
+	ELSE
+		DELETE FROM dbo.ExtendedQueryTagError
+		WHERE TagKey = @tagKey
+
+	COMMIT TRANSACTION
+GO
+
+/***************************************************************************************/
+-- STORED PROCEDURE
+--     AddExtendedQueryTagError
+--
+-- DESCRIPTION
+--    Add an Extended Query Tag Error
+--
+-- PARAMETERS
+--     @tagKey
+--         * The related extended query tag's key
+--     @createdTime
+--         * The time the error was created
+--     @errorCode
+--         * The error code
+--     @studyInstanceUid
+--         * study instance uid
+--     @seriesInstanceUid
+--         * series instance uid
+--     @sopInstanceUid
+--         * sop instance uid
+--     @sopInstanceKey
+--         * sop instance key
+/***************************************************************************************/
+
+CREATE OR ALTER PROCEDURE dbo.AddExtendedQueryTagError (
+	@tagKey INT,
+	@createdTime DATETIME2(7),
+	@errorCode INT,
+	@studyInstanceUid VARCHAR(64),
+	@seriesInstanceUid VARCHAR(64) = NULL,
+	@sopInstanceUid VARCHAR(64) = NULL,
+	@sopInstanceKey BIGINT
+)
+AS
+	SET NOCOUNT     ON
+	SET XACT_ABORT  ON
+	BEGIN TRANSACTION
+
+	-- Check if tag exists
+	IF NOT EXISTS (SELECT * FROM dbo.ExtendedQueryTag WHERE TagKey = @tagKey)
+		THROW 50404, 'Tag does not exist', 1;
+	-- Check if error with same @sopInstanceKey already exist
+	IF EXISTS (SELECT * FROM dbo.ExtendedQueryTagError WHERE sopInstanceKey = @sopInstanceKey)
```

Comment:
What if multiple activities are mistakenly re-indexing the same instance and find the same error? Should we allow it to override?

Reply:
decided to have the XQT error time updated if the tag error already exists and to log the error and not make the system fail and throw